File: dungeons-26.1.2/src/main/java/fishmod/features/BossBarFeature.java
package fishmod.features;

import fishmod.mixin.accessors.BossBarHudAccessor;
import fishmod.utils.debug.Debug;
import fishmod.utils.rendering.RenderUtils;
import java.util.Map;
import java.util.UUID;
import net.minecraft.ChatFormatting;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.LerpingBossEvent;
import net.minecraft.network.chat.Component;

public class BossBarFeature {

    /** Called from HudRenderCallback — fires BEFORE vanilla boss bar, so only used for non-boss-bar elements. */
    public static void renderHud(GuiGraphicsExtractor ctx) {
        // intentionally empty — boss HP drawn in renderAfterVanilla
    }

    /** Called from FishBossBarHudMixin @Inject(RETURN) — fires after vanilla draws its text. */
    public static void renderAfterVanilla(GuiGraphicsExtractor ctx) {
        Minecraft mc = Minecraft.getInstance();
        if (mc == null || mc.player == null) return;

        BossBarHudAccessor accessor = (BossBarHudAccessor) mc.gui.getBossOverlay();
        Map<UUID, LerpingBossEvent> bossBars = accessor.getBossBars();
        if (bossBars == null || bossBars.isEmpty()) return;

        int screenWidth = ctx.guiWidth();
        int y = 12;

        for (LerpingBossEvent bar : bossBars.values()) {
            Component customText = buildText(bar);
            if (customText != null) {
                int textWidth = mc.font.width(customText);
                int textX = screenWidth / 2 - textWidth / 2;
                ctx.text(mc.font, customText, textX, y - 9, 0xFFFFFF, true);
            }
            y += 19;
        }
    }

    private static Component buildText(LerpingBossEvent bar) {
        try {
            String name = bar.getName().getString().replaceAll("§.", "").trim();
            float pct = bar.getProgress() * 100f;
            String pctStr = pct >= 10 ? String.format("%.1f%%", pct) : String.format("%.2f%%", pct);

            float maxHp;
            if (name.contains("Maxor"))       maxHp = 2.5e8f;
            else if (name.contains("Storm"))  maxHp = 5e8f;
            else if (name.contains("Goldor")) maxHp = 7.5e8f;
            else if (name.contains("Necron")) maxHp = 1e9f;
            else maxHp = -1f;

            if (maxHp < 0) {
                return Component.literal(name + " ").withStyle(ChatFormatting.RED)
                        .append(Component.literal(pctStr).withStyle(ChatFormatting.GREEN));
            }

            float currHp = maxHp * bar.getProgress();
            return Component.literal(name + " ").withStyle(ChatFormatting.RED)
                    .append(Component.literal(RenderUtils.formatNumber(currHp)).withStyle(ChatFormatting.GREEN))
                    .append(Component.literal("/").withStyle(ChatFormatting.GRAY))
                    .append(Component.literal(RenderUtils.formatNumber(maxHp)).withStyle(ChatFormatting.GREEN));
        } catch (Exception e) {
            Debug.LOGGER.error("BossBarFeature buildText error: {}", e.getMessage());
            return null;
        }
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/BossBarFeature.kt
package fishmod.features

import fishmod.mixin.accessors.LerpingBossEventAccessor
import fishmod.utils.Location
import fishmod.utils.config.values.Dungeons
import fishmod.utils.dungeon.Phase
import net.minecraft.client.gui.components.LerpingBossEvent
import net.minecraft.network.chat.Component
import java.util.Locale
import kotlin.math.roundToInt

object BossBarFeature {

    private var lastName: Component? = null
    private var lastPercent = Float.NaN
    private var lastOut: Component? = null

    @JvmStatic
    fun appendHealth(instance: LerpingBossEvent, name: Component): Component {
        if (!Dungeons.bossHealthNumbers || !Location.inDungeon()) return name
        val percent = (instance as LerpingBossEventAccessor).targetPercent
        if (name === lastName && percent == lastPercent) lastOut?.let { return it }
        val maxHealth = getMaxHealth(name) ?: return name
        val currentHealth = (percent * maxHealth).roundToInt().toFloat()

        val out = name.copy().append(
            Component.literal(" §r§8- §a${formatHealth(currentHealth)}§7/§a${formatHealth(maxHealth)}§c❤")
        )
        lastName = name
        lastPercent = percent
        lastOut = out
        return out
    }

    private fun getMaxHealth(nameComponent: Component): Float? {
        val name = nameComponent.string.replace(fishmod.utils.Constants.STRIP_COLOR_REGEX, "").trim()
        val floor = Phase.getFloor()
        val master = floor?.startsWith("M", ignoreCase = true) == true
        val floorNum = floor?.filter { it.isDigit() }?.toFloatOrNull() ?: 0f

        return when (name) {
            "The Watcher" -> 12f + floorNum
            "Thorn" -> if (master) 6f else 4f
            "Maxor" -> if (master) 800_000_000f else 100_000_000f
            "Storm" -> if (master) 1_000_000_000f else 400_000_000f
            "Goldor" -> if (master) 1_200_000_000f else 750_000_000f
            "Necron" -> if (master) 1_400_000_000f else 1_000_000_000f
            else -> null
        }
    }

    private fun formatHealth(health: Float): String = when {
        health >= 1_000_000_000 -> {
            val h = health / 1_000_000_000f
            if (h % 1f == 0f) "${h.toInt()}B" else String.format(Locale.US, "%.1fB", h)
        }
        health >= 1_000_000 -> "${(health / 1_000_000f).toInt()}M"
        health >= 1000 -> "${(health / 1000f).toInt()}k"
        else -> health.toInt().toString()
    }
}
